src/client/resources/web/js/chat.mjs

```diff
@@ -4,98 +4,418 @@
 import { faviconCounter, formatTimestamp } from './util.mjs';
 import { assertIsComponent, ComponentError, formatComponent, initializeObfuscation } from './message_parsing.mjs';
 import { parseModServerMessage } from './message_types.mjs';
-/** @typedef {import('./message_parsing.mjs').Component} Component */
 
+/**
+ * Import all types we might need
+ * @typedef {import('./message_parsing.mjs').Component} Component
+ * @typedef {import('./message_types.mjs').ChatMessage} ChatMessage
+ * @typedef {import('./message_types.mjs').HistoryMetaData} HistoryMetaData
+ * @typedef {import('./message_types.mjs').ServerConnectionState} ServerConnectionState
+ */
+
+/**
+ * ======================
+ *  Constants & Globals
+ * ======================
+ */
+
+// WebSocket Management
+/** @type {number} */
+const maxReconnectAttempts = 300; // TODO: add a reconnect button after automatic retries are done.
 /** @type {WebSocket | null} */
 let ws = null;
+/** @type {number} */
 let reconnectAttempts = 0;
-const maxReconnectAttempts = 300; // TODO: add a reconnect button after automatic retries are done.
+
+// Message History Management
+const messageHistoryLimit = 50;
+let isLoadingHistory = false;
 
 // Used for the favicon
 let messageCount = 0;
 
+// Used to keep track of messages already shown. To prevent possible duplication on server join.
+/** @type {Set<string>} */
+const displayedMessageIds = new Set();
+
+/**
+ * Server information and related methods.
+ * @type {{
+ *   name: string | undefined;
+ *   id: string | undefined;
+ *   baseTitle: string;
+ *   update: (name: string , id: string) => void;
+ *   clear: () => void;
+ *   getId: () => string | undefined;
+ *   getName: () => string | undefined;
+ * }}
+ */
+const serverInfo = {
```

Comment:
I would have done this as a class, but it's fine as is.

Reply:
Considered a class, but that did seem a bit overkill as we aren't going to use it elsewhere but still requiring a constructor and all that.